diff --git a/drivers/net/wireless/intel/iwlwifi/iwl-nvm-parse.c b/drivers/net/wireless/intel/iwlwifi/iwl-nvm-parse.c
index 52fe97481..98f1edb72 100644
--- a/drivers/net/wireless/intel/iwlwifi/iwl-nvm-parse.c
+++ b/drivers/net/wireless/intel/iwlwifi/iwl-nvm-parse.c
@@ -1471,7 +1471,9 @@ iwl_parse_nvm_mcc_info(struct device *dev, const struct iwl_cfg *cfg,
 		    band == NL80211_BAND_2GHZ)
 			continue;
 
+#if LINUX_VERSION_IS_GEQ(4,19,0)
 		reg_query_regdb_wmm(regd->alpha2, center_freq, rule);
+#endif
 	}
 
 	regd->n_reg_rules = valid_rules;
diff --git a/drivers/net/wireless/intel/iwlwifi/iwl-tm-gnl.c b/drivers/net/wireless/intel/iwlwifi/iwl-tm-gnl.c
index 55358f5e7..abc500df6 100644
--- a/drivers/net/wireless/intel/iwlwifi/iwl-tm-gnl.c
+++ b/drivers/net/wireless/intel/iwlwifi/iwl-tm-gnl.c
@@ -1084,6 +1084,9 @@ static __genl_const struct genl_ops iwl_tm_gnl_ops[] = {
 #endif
 
 	  .doit = iwl_tm_gnl_cmd_do,
+#if CFG80211_VERSION < KERNEL_VERSION(5,2,0)
+	  .policy = iwl_tm_gnl_msg_policy,
+#endif
 	  .dumpit = iwl_tm_gnl_dump,
 	  .done = iwl_tm_gnl_done,
 	},
@@ -1094,6 +1097,9 @@ static __genl_const struct genl_ops iwl_tm_gnl_ops[] = {
 #endif
 
 		.doit = iwl_tm_gnl_cmd_subscribe,
+#if CFG80211_VERSION < KERNEL_VERSION(5,2,0)
+		.policy	= iwl_tm_gnl_msg_policy,
+#endif
 	},
 };
 
@@ -1102,7 +1108,9 @@ static struct genl_family iwl_tm_gnl_family __genl_ro_after_init = {
 	.name		= IWL_TM_GNL_FAMILY_NAME,
 	.version	= IWL_TM_GNL_VERSION_NR,
 	.maxattr	= IWL_TM_GNL_MSG_ATTR_MAX,
+#if CFG80211_VERSION >= KERNEL_VERSION(5,2,0)
 	.policy		= iwl_tm_gnl_msg_policy,
+#endif
 	.module		= THIS_MODULE,
 	.ops		= iwl_tm_gnl_ops,
 	.n_ops		= ARRAY_SIZE(iwl_tm_gnl_ops),
diff --git a/drivers/net/wireless/intel/iwlwifi/mvm/mac80211.c b/drivers/net/wireless/intel/iwlwifi/mvm/mac80211.c
index bf94e1dcf..592453577 100644
--- a/drivers/net/wireless/intel/iwlwifi/mvm/mac80211.c
+++ b/drivers/net/wireless/intel/iwlwifi/mvm/mac80211.c
@@ -91,6 +91,37 @@
 #endif
 #include "fw/api/nan.h"
 
+/* The ETSI patches were introduced in 4.17 and backported to
+ * chromeos-4.4, but we use our version in 4.4 anyway.  Also, due to
+ * backporting issues (some wmm rules code was moved to cfg80211), we
+ * only use the outer kernel version starting from 4.19.
+ */
+#if CFG80211_VERSION < KERNEL_VERSION(4,19,0)
+const static struct ieee80211_wmm_rule wmm_rules = {
+	.client = {
+		{.cw_min = 3, .cw_max = 7, .aifsn = 2, .cot = 2000},
+		{.cw_min = 7, .cw_max = 15, .aifsn = 2, .cot = 4000},
+		{.cw_min = 15, .cw_max = 1023, .aifsn = 3, .cot = 6000},
+		{.cw_min = 15, .cw_max = 1023, .aifsn = 7, .cot = 6000},
+		},
+	.ap = {
+		{.cw_min = 3, .cw_max = 7, .aifsn = 1, .cot = 2000},
+		{.cw_min = 7, .cw_max = 15, .aifsn = 1, .cot = 4000},
+		{.cw_min = 15, .cw_max = 63, .aifsn = 3, .cot = 6000},
+		{.cw_min = 15, .cw_max = 1023, .aifsn = 7, .cot = 6000},
+	}
+};
+
+const static char *wmm_cc_list[] = {
+	"AT", "BE", "BA", "BG", "HR", "CY", "CZ", "DK", "EE", "FI", "FR", "DE",
+	"GR", "HU", "IS", "IE", "IT", "LV", "LI", "LT", "LU", "MK", "MT", "ME",
+	"NL", "NO", "PL", "PT", "RO", "SK", "SI", "ES", "SE", "CH", "GB", "GF",
+	"PF", "GP", "MQ", "YT", "MC", "NC", "RE", "MF", "PM", "WF", "SM", "VA",
+	"CW", "BQ", "SX", "GL", "AD", "AI", "FK", "JE", "GI", "GG", "MS", "SH",
+	"TC", "VG",
+};
+#endif
+
 static const struct ieee80211_iface_limit iwl_mvm_limits[] = {
 	{
 		.max = CPTCFG_IWLWIFI_NUM_STA_INTERFACES,
@@ -221,6 +252,7 @@ static const struct iwl_fw_bcast_filter iwl_mvm_default_bcast_filters[] = {
 };
 #endif
 
+#if CFG80211_VERSION >= KERNEL_VERSION(5,0,0)
 static const struct cfg80211_pmsr_capabilities iwl_mvm_pmsr_capa = {
 	.max_peers = IWL_MVM_TOF_MAX_APS,
 	.report_ap_tsf = 1,
@@ -246,6 +278,7 @@ static const struct cfg80211_pmsr_capabilities iwl_mvm_pmsr_capa = {
 			     BIT(NL80211_PREAMBLE_HE),
 	},
 };
+#endif
 
 static int __iwl_mvm_mac_set_key(struct ieee80211_hw *hw,
 				 enum set_key_cmd cmd,
@@ -561,12 +593,14 @@ int iwl_mvm_mac_setup_register(struct iwl_mvm *mvm)
 		hw->wiphy->n_cipher_suites++;
 	}
 
+#if CFG80211_VERSION >= KERNEL_VERSION(5,0,0)
 	if (fw_has_capa(&mvm->fw->ucode_capa,
 			IWL_UCODE_TLV_CAPA_FTM_CALIBRATED)) {
 		wiphy_ext_feature_set(hw->wiphy,
 				      NL80211_EXT_FEATURE_ENABLE_FTM_RESPONDER);
 		hw->wiphy->pmsr_capa = &iwl_mvm_pmsr_capa;
 	}
+#endif
 
 	if (fw_has_capa(&mvm->fw->ucode_capa,
 			IWL_UCODE_TLV_CAPA_BIGTK_SUPPORT))
@@ -689,6 +723,7 @@ int iwl_mvm_mac_setup_register(struct iwl_mvm *mvm)
 	/* we create the 802.11 header and zero length SSID IE. */
 	hw->wiphy->max_sched_scan_ie_len =
 		SCAN_OFFLOAD_PROBE_REQ_SIZE - 24 - 2;
+#if CFG80211_VERSION >= KERNEL_VERSION(4,4,0)
 	hw->wiphy->max_sched_scan_plans = IWL_MAX_SCHED_SCAN_PLANS;
 	hw->wiphy->max_sched_scan_plan_interval = U16_MAX;
 
@@ -697,6 +732,7 @@ int iwl_mvm_mac_setup_register(struct iwl_mvm *mvm)
 	 * infinite loop, so the maximum number of iterations is actually 254.
 	 */
 	hw->wiphy->max_sched_scan_plan_iterations = 254;
+#endif
 
 	hw->wiphy->features |= NL80211_FEATURE_P2P_GO_CTWIN |
 			       NL80211_FEATURE_LOW_PRIORITY_SCAN |
@@ -3425,6 +3465,66 @@ static void iwl_mvm_sta_rc_update(struct ieee80211_hw *hw,
 		iwl_mvm_sf_update(mvm, vif, false);
 }
 
+#if CFG80211_VERSION < KERNEL_VERSION(4,19,0)
+static void iwl_mvm_limit_wmm_ac(struct iwl_mvm *mvm,
+				 struct ieee80211_vif *vif,
+				 struct ieee80211_tx_queue_params *params,
+				 u16 ac)
+{
+	struct ieee80211_regdomain *rd;
+	struct ieee80211_chanctx_conf *chanctx_conf;
+	const struct ieee80211_wmm_ac *wmm_ac;
+	u16 center_freq = 0;
+	int i;
+
+	rcu_read_lock();
+	chanctx_conf = rcu_dereference(vif->chanctx_conf);
+	if (chanctx_conf)
+		center_freq = chanctx_conf->def.chan->center_freq;
+
+	rcu_read_unlock();
+
+	if (!center_freq || center_freq < 5180 || center_freq > 5720)
+		return;
+
+	if (vif->type != NL80211_IFTYPE_STATION &&
+	    vif->type != NL80211_IFTYPE_AP &&
+	    vif->type != NL80211_IFTYPE_P2P_CLIENT &&
+	    vif->type != NL80211_IFTYPE_P2P_GO)
+		return;
+
+	mutex_lock(&mvm->mutex);
+	if (iwl_mvm_is_lar_supported(mvm))
+		rd = iwl_mvm_get_current_regdomain(mvm, NULL);
+	else
+		rd = NULL;
+	mutex_unlock(&mvm->mutex);
+
+	if (IS_ERR_OR_NULL(rd))
+		return;
+
+	for  (i = 0; i < ARRAY_SIZE(wmm_cc_list); i++) {
+		if (!strncmp(wmm_cc_list[i], rd->alpha2, 2)) {
+			if (vif->type == NL80211_IFTYPE_STATION ||
+			    vif->type == NL80211_IFTYPE_P2P_CLIENT)
+				wmm_ac = &wmm_rules.client[ac];
+			else
+				wmm_ac = &wmm_rules.ap[ac];
+
+			params->txop =
+				min_t(u16, params->txop, wmm_ac->cot / 32);
+			params->cw_min =
+				max_t(u16, params->cw_min, wmm_ac->cw_min);
+			params->cw_max =
+				max_t(u16, params->cw_max, wmm_ac->cw_max);
+			params->aifs =
+				max_t(u8, params->aifs, wmm_ac->aifsn);
+			return;
+		}
+	}
+}
+#endif
+
 static int iwl_mvm_mac_conf_tx(struct ieee80211_hw *hw,
 			       struct ieee80211_vif *vif, u16 ac,
 			       const struct ieee80211_tx_queue_params *params)
@@ -3434,6 +3534,10 @@ static int iwl_mvm_mac_conf_tx(struct ieee80211_hw *hw,
 
 	mvmvif->queue_params[ac] = *params;
 
+#if CFG80211_VERSION < KERNEL_VERSION(4,19,0)
+	iwl_mvm_limit_wmm_ac(mvm, vif, &mvmvif->queue_params[ac], ac);
+#endif
+
 	/*
 	 * No need to update right away, we'll get BSS_CHANGED_QOS
 	 * The exception is P2P_DEVICE interface which needs immediate update.
@@ -5004,6 +5108,7 @@ static void iwl_mvm_set_sta_rate(u32 rate_n_flags, struct rate_info *rinfo)
 		if (rate_n_flags & RATE_MCS_SGI_MSK)
 			rinfo->flags |= RATE_INFO_FLAGS_SHORT_GI;
 	} else if (rate_n_flags & RATE_MCS_HE_MSK) {
+#if CFG80211_VERSION >= KERNEL_VERSION(4,19,0)
 		u32 gi_ltf = u32_get_bits(rate_n_flags,
 					  RATE_MCS_HE_GI_LTF_MSK);
 
@@ -5048,6 +5153,7 @@ static void iwl_mvm_set_sta_rate(u32 rate_n_flags, struct rate_info *rinfo)
 
 		if (rate_n_flags & RATE_HE_DUAL_CARRIER_MODE_MSK)
 			rinfo->he_dcm = 1;
+#endif
 	} else {
 		switch (u32_get_bits(rate_n_flags, RATE_LEGACY_RATE_MSK)) {
 		case IWL_RATE_1M_PLCP:
diff --git a/drivers/net/wireless/intel/iwlwifi/mvm/scan.c b/drivers/net/wireless/intel/iwlwifi/mvm/scan.c
index 49159a2b2..cbd63694b 100644
--- a/drivers/net/wireless/intel/iwlwifi/mvm/scan.c
+++ b/drivers/net/wireless/intel/iwlwifi/mvm/scan.c
@@ -2265,8 +2265,10 @@ int iwl_mvm_reg_scan_start(struct iwl_mvm *mvm, struct ieee80211_vif *vif,
 
 	iwl_mvm_fill_scan_type(mvm, &params, vif);
 
+#if CFG80211_VERSION >= KERNEL_VERSION(4,9,0)
 	if (req->duration)
 		params.iter_notif = true;
+#endif
 
 	iwl_mvm_build_scan_probe(mvm, vif, ies, &params);
 
@@ -2663,11 +2663,20 @@ int iwl_mvm_sched_scan_start(struct iwl_mvm *mvm,
 	params.pass_all =  iwl_mvm_scan_pass_all(mvm, req);
 	params.n_match_sets = req->n_match_sets;
 	params.match_sets = req->match_sets;
+#if CFG80211_VERSION >= KERNEL_VERSION(4,4,0)
 	if (!req->n_scan_plans)
 		return -EINVAL;
 
 	params.n_scan_plans = req->n_scan_plans;
 	params.scan_plans = req->scan_plans;
+#else
+	params.n_scan_plans = 1;
+	params.scan_plans = &scan_plan;
+	if (req->interval / MSEC_PER_SEC > U16_MAX)
+		scan_plan.interval = U16_MAX;
+	else
+		scan_plan.interval = req->interval / MSEC_PER_SEC;
+#endif
 
 	iwl_mvm_fill_scan_type(mvm, &params, vif);
 
diff --git a/drivers/net/wireless/intel/iwlwifi/pcie/drv.c b/drivers/net/wireless/intel/iwlwifi/pcie/drv.c
index 79caefcc2..437ad0fe1 100644
--- a/drivers/net/wireless/intel/iwlwifi/pcie/drv.c
+++ b/drivers/net/wireless/intel/iwlwifi/pcie/drv.c
@@ -60,6 +60,7 @@
  * Copyright (C) 2013-2015 Intel Mobile Communications GmbH
  * Copyright (C) 2016-2017 Intel Deutschland GmbH
  */
+#undef pr_fmt
 #define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
 
 #include <linux/module.h>
